src/core/or/circuit_st.h

```diff
@@ -177,6 +182,19 @@ struct circuit_t {
   /** Hashtable node: used to look up the circuit by its HS token using the HS
       circuitmap. */
   HT_ENTRY(circuit_t) hs_circuitmap_node;
+
+  /** Adaptive Padding state machines: these are immutable. The state machines
+   *  that come from the consensus are saved to a global structure, to avoid
+   *  per-circuit allocations. This merely points to the global copy. */
+  const circpad_machine_t *padding_machine[CIRCPAD_MAX_MACHINES];
+
+  /** Adaptive Padding machine info for above machines. This is the
+   *  per-circuit mutable information, such as the current state and
+   *  histogram token counts. Some of it is optional (aka NULL).
+   *  If a machine is being shut down, these indexes can be NULL
+   *  without the corresponding padding_machine being NULL, while we
+   *  wait for the other end to respond to our shutdown request. */
+  circpad_machineinfo_t *padding_info[CIRCPAD_MAX_MACHINES];
```

Comment:
As above; what do the indices mean?

Reply:
Addressed in 53fb50bd3cfccd93a3d17af18f2f9c0ad42a7cf9.